File: backend/app/schemas/telemetry.py
"""Pydantic schemas for data source telemetry, freshness, and ingestion run API endpoints."""

from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field

from app.core.telemetry.contracts import FreshnessStatus


class FreshnessEvaluationRead(BaseModel):
    """Schema representing evaluated freshness and usability."""

    model_config = ConfigDict(from_attributes=True)

    status: FreshnessStatus
    age_seconds: Optional[float] = None
    threshold_seconds: float
    is_usable: bool
    reason: str
    evaluated_at: datetime


class DataIngestionRunRead(BaseModel):
    """Schema for an individual ingestion run record."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    data_source_id: int
    status: str
    records_ingested: int
    records_failed: int
    started_at: datetime
    completed_at: Optional[datetime] = None
    log_details: Optional[str] = None


class DataSourceTelemetryRead(BaseModel):
    """Schema representing operational telemetry for a data source."""

    model_config = ConfigDict(from_attributes=True)

    source_id: int
    name: str
    source_type: str
    provider: str
    provider_id: Optional[str] = None
    category: Optional[str] = None
    provider_mode: str
    provider_health: str
    freshness: FreshnessEvaluationRead
    last_successful_update: Optional[datetime] = None
    last_attempted_update: Optional[datetime] = None
    latest_run_status: Optional[str] = None
    is_synthetic: bool = True
    is_active: bool = True
    endpoint_url: Optional[str] = None
    polling_interval_seconds: Optional[int] = None
    region_id: Optional[str] = None
    records_ingested_total: int = 0
    records_failed_total: int = 0


class DataSourceDetailRead(DataSourceTelemetryRead):
    """Detailed data source telemetry with recent ingestion runs and metadata."""

    recent_runs: List[DataIngestionRunRead] = Field(default_factory=list)
    metadata_json: Optional[Dict[str, Any]] = None


class TelemetryOverviewRead(BaseModel):
    """Aggregated platform-wide data source health and freshness overview."""

    model_config = ConfigDict(from_attributes=True)

    total_sources: int
    healthy_count: int
    degraded_count: int
    unavailable_count: int
    fresh_count: int
    stale_count: int
    unknown_count: int
    synthetic_count: int
    evaluated_at: datetime

File: frontend/src/types/telemetry.ts
/**
 * Strongly typed domain models for Data Sources & Freshness Monitoring (Chunk M6-06).
 * Conforms strictly to:
 * - Chunk M3-13 Data Source Freshness & Telemetry Backend (contracts.py, telemetry.py)
 * - Chunk M3-03 Provider Interfaces & Mock Adapters (contracts.py)
 */

// ============================================================================
// Enums & Literal Unions (strictly mirroring M3-13 & M3-03)
// ============================================================================

export type FreshnessStatus =
  | "fresh"
  | "stale"
  | "unavailable"
  | "clock_skew"
  | "unknown";

export type ProviderHealth =
  | "healthy"
  | "degraded"
  | "unavailable"
  | "unknown";

export type ProviderMode =
  | "live"
  | "mock"
  | "file"
  | "hybrid";

export type SourceCategory =
  | "rainfall"
  | "flood"
  | "landslide"
  | "hazard_observation"
  | "population_exposure"
  | "other";

// ============================================================================
// Evaluation & Telemetry Models
// ============================================================================

export interface FreshnessEvaluationRead {
  status: FreshnessStatus;
  age_seconds: number | null;
  threshold_seconds: number;
  is_usable: boolean;
  reason: string;
  evaluated_at: string;
}

export interface DataIngestionRunRead {
  id: number;
  data_source_id: number;
  status: "success" | "failed" | "running" | "partial" | string;
  records_ingested: number;
  records_failed: number;
  started_at: string;
  completed_at: string | null;
  log_details: string | null;
}

export interface DataSourceTelemetryRead {
  source_id: number;
  name: string;
  source_type: string;
  provider: string;
  provider_id: string | null;
  category: SourceCategory | string | null;
  provider_mode: ProviderMode | string;
  provider_health: ProviderHealth | string;
  freshness: FreshnessEvaluationRead;
  last_successful_update: string | null;
  last_attempted_update: string | null;
  latest_run_status: string | null;
  is_synthetic: boolean;
  is_active: boolean;
  endpoint_url: string | null;
  polling_interval_seconds: number | null;
  region_id: string | null;
  records_ingested_total: number;
  records_failed_total: number;
}

export interface DataSourceDetailRead extends DataSourceTelemetryRead {
  recent_runs: DataIngestionRunRead[];
  metadata_json: Record<string, any> | null;
}

export interface TelemetryOverviewRead {
  total_sources: number;
  healthy_count: number;
  degraded_count: number;
  unavailable_count: number;
  fresh_count: number;
  stale_count: number;
  unknown_count: number;
  synthetic_count: number;
  evaluated_at: string;
}

export interface CategoryFreshnessThresholdItem {
  category: SourceCategory;
  name: string;
  threshold_seconds: number;
  description: string;
}

export interface DataSourceFilterCriteria {
  search?: string;
  category?: string;
  health?: string;
  freshness?: string;
  mode?: string;
}
